File: rust/katari-runtime/src/runtime/execute.rs
use crate::ir::instruction::Instruction;
use crate::ir::{ConstVal, IRModule, ThreadKind};
use crate::value::{self, Value};

use super::agent::{AgentState, AgentStatus};
use super::event::{Event, EventKind};
use super::signal::Signal;
use super::thread::{ThreadState, ThreadStatus};
use super::{handle, for_loop, par, request};

// ---------------------------------------------------------------------------
// Thread lifecycle helpers
// ---------------------------------------------------------------------------

/// Complete a thread with the given signal.
/// Root thread: store result in AgentStatus, remove thread, push AgentCompleted.
/// Non-root: remove thread, push ThreadCompleted to parent.
pub fn finish_thread(
    agent: &mut AgentState,
    thread_id: u32,
    signal: Signal,
    events: &mut Vec<Event>,
) {
    if thread_id == agent.root_thread {
        agent.status = match &signal {
            Signal::Normal(v) | Signal::FnReturn(v) => AgentStatus::Completed(v.clone()),
            _ => AgentStatus::Error,
        };
        agent.threads.remove(&thread_id);
        events.push(Event {
            agent_id: agent.agent_id.clone(),
            kind: EventKind::AgentCompleted,
        });
    } else {
        let (kind, parent) = match agent.threads.get(&thread_id) {
            Some(t) => (t.kind, t.parent),
            None => return,
        };
        agent.threads.remove(&thread_id);
        if let Some(parent_id) = parent {
            events.push(Event {
                agent_id: agent.agent_id.clone(),
                kind: EventKind::ThreadCompleted {
                    parent_id,
                    child_id: thread_id,
                    child_kind: kind,
                    signal,
                },
            });
        }
    }
}

/// Resume a suspended thread and push an Execute event.
pub fn resume_thread(
    agent: &mut AgentState,
    thread_id: u32,
    events: &mut Vec<Event>,
) {
    if let Some(t) = agent.threads.get_mut(&thread_id) {
        t.status = ThreadStatus::Running;
    }
    events.push(Event {
        agent_id: agent.agent_id.clone(),
        kind: EventKind::Execute(thread_id),
    });
}

/// Create a new child thread and push an Execute event.
pub fn spawn_child_thread(
    agent: &mut AgentState,
    tid: u32,
    kind: ThreadKind,
    parent: u32,
    events: &mut Vec<Event>,
) {
    let thread = ThreadState::new(tid, kind, Some(parent));
    agent.threads.insert(tid, thread);
    events.push(Event {
        agent_id: agent.agent_id.clone(),
        kind: EventKind::Execute(tid),
    });
}

// ---------------------------------------------------------------------------
// Thread execution
// ---------------------------------------------------------------------------

/// Execute a thread's instructions until it suspends or completes.
pub fn execute_thread(
    agent: &mut AgentState,
    module: &IRModule,
    thread_id: u32,
    events: &mut Vec<Event>,
) {
    let ir_thread = match module.threads.iter().find(|t| t.id == thread_id) {
        Some(t) => t,
        None => {
            finish_thread(agent, thread_id, Signal::Normal(Value::Null), events);
            return;
        }
    };

    loop {
        let pc = match agent.threads.get(&thread_id) {
            Some(t) if t.is_running() => t.pc as usize,
            _ => return,
        };

        if pc >= ir_thread.body.len() {
            finish_thread(agent, thread_id, Signal::Normal(Value::Null), events);
            return;
        }

        let instr = ir_thread.body[pc].clone();

        if let Some(t) = agent.threads.get_mut(&thread_id) {
            t.pc += 1;
        }

        match instr {
            // === Terminal instructions ===
            Instruction::Complete(val) => {
                let v = agent.get_var(val);
                finish_thread(agent, thread_id, Signal::Normal(v), events);
                return;
            }
            Instruction::Return(val) => {
                let v = agent.get_var(val);
                finish_thread(agent, thread_id, Signal::FnReturn(v), events);
                return;
            }
            Instruction::HandleBreak(val) => {
                let v = agent.get_var(val);
                finish_thread(agent, thread_id, Signal::HandleBreak(v), events);
                return;
            }
            Instruction::Continue(val, mutations) => {
                let v = agent.get_var(val);
                finish_thread(agent, thread_id, Signal::Continue(v, mutations), events);
                return;
            }
            Instruction::ForBreak(val) => {
                let v = agent.get_var(val);
                finish_thread(agent, thread_id, Signal::ForBreak(v), events);
                return;
            }
            Instruction::ForContinue(mutations) => {
                finish_thread(agent, thread_id, Signal::ForContinue(mutations), events);
                return;
            }

            // === Control flow ===
            Instruction::Jump(target) => {
                if let Some(t) = agent.threads.get_mut(&thread_id) {
                    t.pc = target;
                }
            }
            Instruction::Branch(cond, then_target, else_target) => {
                let v = agent.get_var(cond);
                let target = if v.is_truthy() { then_target } else { else_target };
                if let Some(t) = agent.threads.get_mut(&thread_id) {
                    t.pc = target;
                }
            }
            Instruction::Switch(val, cases, default) => {
                let v = agent.get_var(val);
                let target = find_switch_target(&v, &cases, default, &module.consts);
                if let Some(t) = agent.threads.get_mut(&thread_id) {
                    t.pc = target;
                }
            }

            // === Suspension points ===
            Instruction::Handle(dst, hid) => {
                handle::setup(agent, module, thread_id, dst, hid, events);
                return;
            }
            Instruction::For(dst, fid) => {
                for_loop::setup(agent, module, thread_id, dst, fid, events);
                return;
            }
            Instruction::Par(dst, tids) => {
                par::setup(agent, thread_id, dst, &tids, events);
                return;
            }
            Instruction::Call(dst, aid, args) => {
                request::handle_icall(agent, module, thread_id, dst, aid, &args, events);
                if !agent
                    .threads
                    .get(&thread_id)
                    .is_some_and(|t| t.is_running())
                {
                    return;
                }
            }
            Instruction::Request(dst, rid, args) => {
                request::handle_irequest(agent, module, thread_id, dst, rid, &args, events);
                return;
            }

            // === Constants & movement ===
            Instruction::LoadConst(dst, cid) => {
                let val = load_const(&module.consts, cid);
                agent.set_var(dst, val);
            }
            Instruction::LoadNull(dst) => {
                agent.set_var(dst, Value::Null);
            }
            Instruction::Move(dst, src) => {
                let v = agent.get_var(src);
                agent.set_var(dst, v);
            }

            // === Object operations ===
            Instruction::NewObject(dst, fields) => {
                let mut obj = indexmap::IndexMap::new();
                for (cid, vid) in fields {
                    let key = const_as_string(&module.consts, cid);
                    let val = agent.get_var(vid);
                    obj.insert(key, val);
                }
                agent.set_var(dst, Value::Object(obj));
            }
            Instruction::GetField(dst, obj, field_cid) => {
                let o = agent.get_var(obj);
                let key = const_as_string(&module.consts, field_cid);
                let val = o
                    .as_object()
                    .and_then(|m| m.get(&key))
                    .cloned()
                    .unwrap_or(Value::Null);
                agent.set_var(dst, val);
            }
            Instruction::SetField(new_dst, obj, field_cid, val) => {
                let o = agent.get_var(obj);
                let key = const_as_string(&module.consts, field_cid);
                let v = agent.get_var(val);
                let mut new_obj = match o {
                    Value::Object(map) => map,
                    _ => indexmap::IndexMap::new(),
                };
                new_obj.insert(key, v);
                agent.set_var(new_dst, Value::Object(new_obj));
            }
            Instruction::HasField(dst, obj, field_cid) => {
                let o = agent.get_var(obj);
                let key = const_as_string(&module.consts, field_cid);
                let has = o.as_object().is_some_and(|m| m.contains_key(&key));
                agent.set_var(dst, Value::Boolean(has));
            }

            // === Array operations ===
            Instruction::NewArray(dst, elems) => {
                let arr: Vec<Value> = elems.iter().map(|v| agent.get_var(*v)).collect();
                agent.set_var(dst, Value::Array(arr));
            }
            Instruction::ArrGet(dst, arr, idx) => {
                let a = agent.get_var(arr);
                let i = agent.get_var(idx);
                let val = match (a.as_array(), i.as_integer()) {
                    (Some(arr), Some(idx)) => {
                        let idx = if idx < 0 {
                            (arr.len() as i64 + idx) as usize
                        } else {
                            idx as usize
                        };
                        arr.get(idx).cloned().unwrap_or(Value::Null)
                    }
                    _ => Value::Null,
                };
                agent.set_var(dst, val);
            }
            Instruction::ArrLen(dst, arr) => {
                let a = agent.get_var(arr);
                let len = a.as_array().map_or(0, |a| a.len());
                agent.set_var(dst, Value::Integer(len as i64));
            }
            Instruction::ArrPush(dst, arr, elem) => {
                let a = agent.get_var(arr);
                let e = agent.get_var(elem);
                let mut v = match a {
                    Value::Array(arr) => arr,
                    _ => vec![],
                };
                v.push(e);
                agent.set_var(dst, Value::Array(v));
            }
            Instruction::ArrSlice(dst, arr, start, end) => {
                let a = agent.get_var(arr);
                let s = agent.get_var(start);
                let e = agent.get_var(end);
                let val = match (a.as_array(), s.as_integer(), e.as_integer()) {
                    (Some(arr), Some(s), Some(e)) => {
                        let s = (s.max(0) as usize).min(arr.len());
                        let e = (e.max(0) as usize).min(arr.len());
                        if s <= e {
                            Value::Array(arr[s..e].to_vec())
                        } else {
                            Value::Array(vec![])
                        }
                    }
                    _ => Value::Array(vec![]),
                };
                agent.set_var(dst, val);
            }

            // === Arithmetic ===
            Instruction::Add(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::arith_add(&l, &r).unwrap_or(Value::Null));
            }
            Instruction::Sub(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::arith_sub(&l, &r).unwrap_or(Value::Null));
            }
            Instruction::Mul(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::arith_mul(&l, &r).unwrap_or(Value::Null));
            }
            Instruction::Div(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::arith_div(&l, &r).unwrap_or(Value::Null));
            }
            Instruction::Mod(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::arith_mod(&l, &r).unwrap_or(Value::Null));
            }
            Instruction::Neg(dst, src) => {
                let v = agent.get_var(src);
                agent.set_var(dst, value::arith_neg(&v).unwrap_or(Value::Null));
            }

            // === Comparison ===
            Instruction::CmpEq(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::cmp_eq(&l, &r));
            }
            Instruction::CmpNe(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::cmp_ne(&l, &r));
            }
            Instruction::CmpLt(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::cmp_lt(&l, &r).unwrap_or(Value::Boolean(false)));
            }
            Instruction::CmpLe(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::cmp_le(&l, &r).unwrap_or(Value::Boolean(false)));
            }
            Instruction::CmpGt(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::cmp_gt(&l, &r).unwrap_or(Value::Boolean(false)));
            }
            Instruction::CmpGe(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::cmp_ge(&l, &r).unwrap_or(Value::Boolean(false)));
            }

            // === Logical ===
            Instruction::And(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::logic_and(&l, &r).unwrap_or(Value::Boolean(false)));
            }
            Instruction::Or(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::logic_or(&l, &r).unwrap_or(Value::Boolean(false)));
            }
            Instruction::Not(dst, src) => {
                let v = agent.get_var(src);
                agent.set_var(dst, value::logic_not(&v).unwrap_or(Value::Boolean(false)));
            }

            // === String/Type ===
            Instruction::Concat(dst, lhs, rhs) => {
                let (l, r) = (agent.get_var(lhs), agent.get_var(rhs));
                agent.set_var(dst, value::concat(&l, &r).unwrap_or(Value::Null));
            }
            Instruction::ToString(dst, src) => {
                let v = agent.get_var(src);
                agent.set_var(dst, Value::String(v.to_display_string()));
            }
            Instruction::TypeOf(dst, src) => {
                let v = agent.get_var(src);
                agent.set_var(dst, Value::String(v.type_name().to_string()));
            }
        }
    }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

pub fn load_const(consts: &[ConstVal], cid: u32) -> Value {
    match consts.get(cid as usize) {
        Some(ConstVal::Null) => Value::Null,
        Some(ConstVal::Bool(b)) => Value::Boolean(*b),
        Some(ConstVal::Int(n)) => Value::Integer(*n),
        Some(ConstVal::Num(n)) => Value::Number(*n),
        Some(ConstVal::Str(s)) => Value::String(s.clone()),
        None => Value::Null,
    }
}

pub fn const_as_string(consts: &[ConstVal], cid: u32) -> String {
    match consts.get(cid as usize) {
        Some(ConstVal::Str(s)) => s.clone(),
        _ => String::new(),
    }
}

fn find_switch_target(val: &Value, cases: &[(u32, u32)], default: u32, consts: &[ConstVal]) -> u32 {
    for (cid, target) in cases {
        let case_val = load_const(consts, *cid);
        if val == &case_val {
            return *target;
        }
    }
    default
}

File: ts/packages/katari-runtime/src/runtime/execute.ts
import type { AgentState } from "./types.js";
import { getVar, setVar, constAsString } from "./types.js";
import type { Value } from "../value.js";
import {
  constToValue,
  isTruthy,
  toDisplayString,
  typeName,
  valueAdd,
  valueSub,
  valueMul,
  valueDiv,
  valueMod,
  valueNeg,
  valueEq,
  valueLt,
  valueConcat,
} from "../value.js";

// ===========================================================================
// StepResult — what stopped VM execution
// ===========================================================================

export type StepResult =
  // Terminal — produces a runtime event on this thread
  | { tag: "completed"; value: Value }
  | { tag: "returned"; value: Value }
  | { tag: "broken"; value: Value }
  | { tag: "continued"; value: Value; mutations: [number, number][] }
  | { tag: "for_broken"; value: Value }
  | { tag: "for_continued"; mutations: [number, number][] }
  // Suspension — needs orchestration by dispatch
  | { tag: "handle"; dst: number; handleId: number }
  | { tag: "for"; dst: number; forId: number }
  | { tag: "par"; dst: number; threads: number[] }
  | { tag: "call"; dst: number; agentDefId: number; args: Record<string, Value> }
  | { tag: "request"; dst: number; reqDefId: number; args: Record<string, Value> };

// ===========================================================================
// VM: execute thread instructions until a stop point
//
// Pure computation — no side effects beyond mutating agent.scopes.
// Returns what stopped execution, or null if the thread is missing.
// ===========================================================================

export function executeThread(
  agent: AgentState,
  threadId: number
): StepResult | null {
  const thread = agent.threads.get(threadId);
  if (!thread) return null;

  const irThread = agent.module.threads.get(thread.blockId);
  if (!irThread) return { tag: "completed", value: null };

  const s = thread.scopeId;

  for (;;) {
    if (thread.pc >= irThread.body.length) {
      return { tag: "completed", value: null };
    }

    const instr = irThread.body[thread.pc]!;
    thread.pc++;

    switch (instr.op) {
      // --- Terminal ---
      case "Complete":
        return { tag: "completed", value: getVar(agent, s, instr.val) };
      case "Return":
        return { tag: "returned", value: getVar(agent, s, instr.val) };
      case "HandleBreak":
        return { tag: "broken", value: getVar(agent, s, instr.val) };
      case "Continue":
        return {
          tag: "continued",
          value: getVar(agent, s, instr.val),
          mutations: instr.mutations,
        };
      case "ForBreak":
        return { tag: "for_broken", value: getVar(agent, s, instr.val) };
      case "ForContinue":
        return { tag: "for_continued", mutations: instr.mutations };

      // --- Suspension ---
      case "Handle":
        return { tag: "handle", dst: instr.dst, handleId: instr.handleId };
      case "For":
        return { tag: "for", dst: instr.dst, forId: instr.forId };
      case "Par":
        return { tag: "par", dst: instr.dst, threads: instr.threads };
      case "Call": {
        const args: Record<string, Value> = {};
        for (const [name, vid] of instr.args) args[name] = getVar(agent, s, vid);
        return { tag: "call", dst: instr.dst, agentDefId: instr.agentDefId, args };
      }
      case "Request": {
        const args: Record<string, Value> = {};
        for (const [name, vid] of instr.args) args[name] = getVar(agent, s, vid);
        return { tag: "request", dst: instr.dst, reqDefId: instr.reqDefId, args };
      }

      // --- Control flow ---
      case "Jump":
        thread.pc = instr.target;
        break;
      case "Branch":
        thread.pc = isTruthy(getVar(agent, s, instr.cond))
          ? instr.thenPc
          : instr.elsePc;
        break;
      case "Switch": {
        const val = getVar(agent, s, instr.val);
        let target = instr.defaultPc;
        for (const [cid, pc] of instr.cases) {
          if (valueEq(val, constToValue(agent.module.consts[cid]!))) {
            target = pc;
            break;
          }
        }
        thread.pc = target;
        break;
      }

      // --- Constants & Movement ---
      case "LoadConst":
        setVar(
          agent,
          s,
          instr.dst,
          constToValue(agent.module.consts[instr.cid]!)
        );
        break;
      case "LoadNull":
        setVar(agent, s, instr.dst, null);
        break;
      case "Move":
        setVar(agent, s, instr.dst, getVar(agent, s, instr.src));
        break;

      // --- Object ---
      case "NewObject": {
        const obj: Record<string, Value> = {};
        for (const [cid, vid] of instr.fields) {
          obj[constAsString(agent.module.consts, cid)] = getVar(agent, s, vid);
        }
        setVar(agent, s, instr.dst, obj);
        break;
      }
      case "GetField": {
        const o = getVar(agent, s, instr.obj);
        const key = constAsString(agent.module.consts, instr.field);
        setVar(
          agent,
          s,
          instr.dst,
          o && typeof o === "object" && !Array.isArray(o)
            ? (o[key] ?? null)
            : null
        );
        break;
      }
      case "SetField": {
        const o = getVar(agent, s, instr.obj);
        const key = constAsString(agent.module.consts, instr.field);
        const val = getVar(agent, s, instr.val);
        const base =
          o && typeof o === "object" && !Array.isArray(o) ? { ...o } : {};
        base[key] = val;
        setVar(agent, s, instr.dst, base);
        break;
      }
      case "HasField": {
        const o = getVar(agent, s, instr.obj);
        const key = constAsString(agent.module.consts, instr.field);
        setVar(
          agent,
          s,
          instr.dst,
          !!(o && typeof o === "object" && !Array.isArray(o) && key in o)
        );
        break;
      }

      // --- Array ---
      case "NewArray":
        setVar(
          agent,
          s,
          instr.dst,
          instr.elems.map((v) => getVar(agent, s, v))
        );
        break;
      case "ArrGet": {
        const arr = getVar(agent, s, instr.arr);
        const idx = getVar(agent, s, instr.idx);
        if (Array.isArray(arr) && typeof idx === "number") {
          const i = idx < 0 ? arr.length + idx : idx;
          setVar(agent, s, instr.dst, arr[i] ?? null);
        } else {
          setVar(agent, s, instr.dst, null);
        }
        break;
      }
      case "ArrLen": {
        const arr = getVar(agent, s, instr.arr);
        setVar(agent, s, instr.dst, Array.isArray(arr) ? arr.length : 0);
        break;
      }
      case "ArrPush": {
        const arr = getVar(agent, s, instr.arr);
        const elem = getVar(agent, s, instr.elem);
        setVar(
          agent,
          s,
          instr.dst,
          Array.isArray(arr) ? [...arr, elem] : [elem]
        );
        break;
      }
      case "ArrSlice": {
        const arr = getVar(agent, s, instr.arr);
        const start = getVar(agent, s, instr.start);
        const end = getVar(agent, s, instr.end);
        if (
          Array.isArray(arr) &&
          typeof start === "number" &&
          typeof end === "number"
        ) {
          const sl = Math.min(Math.max(0, start), arr.length);
          const e = Math.min(Math.max(0, end), arr.length);
          setVar(agent, s, instr.dst, sl <= e ? arr.slice(sl, e) : []);
        } else {
          setVar(agent, s, instr.dst, []);
        }
        break;
      }

      // --- Arithmetic ---
      case "Add":
        setVar(agent, s, instr.dst, valueAdd(getVar(agent, s, instr.lhs), getVar(agent, s, instr.rhs)));
        break;
      case "Sub":
        setVar(agent, s, instr.dst, valueSub(getVar(agent, s, instr.lhs), getVar(agent, s, instr.rhs)));
        break;
      case "Mul":
        setVar(agent, s, instr.dst, valueMul(getVar(agent, s, instr.lhs), getVar(agent, s, instr.rhs)));
        break;
      case "Div":
        setVar(agent, s, instr.dst, valueDiv(getVar(agent, s, instr.lhs), getVar(agent, s, instr.rhs)));
        break;
      case "Mod":
        setVar(agent, s, instr.dst, valueMod(getVar(agent, s, instr.lhs), getVar(agent, s, instr.rhs)));
        break;
      case "Neg":
        setVar(agent, s, instr.dst, valueNeg(getVar(agent, s, instr.src)));
        break;

      // --- Comparison ---
      case "CmpEq":
        setVar(agent, s, instr.dst, valueEq(getVar(agent, s, instr.lhs), getVar(agent, s, instr.rhs)));
        break;
      case "CmpNe":
        setVar(agent, s, instr.dst, !valueEq(getVar(agent, s, instr.lhs), getVar(agent, s, instr.rhs)));
        break;
      case "CmpLt":
        setVar(agent, s, instr.dst, valueLt(getVar(agent, s, instr.lhs), getVar(agent, s, instr.rhs)));
        break;
      case "CmpLe":
        setVar(agent, s, instr.dst, !valueLt(getVar(agent, s, instr.rhs), getVar(agent, s, instr.lhs)));
        break;
      case "CmpGt":
        setVar(agent, s, instr.dst, valueLt(getVar(agent, s, instr.rhs), getVar(agent, s, instr.lhs)));
        break;
      case "CmpGe":
        setVar(agent, s, instr.dst, !valueLt(getVar(agent, s, instr.lhs), getVar(agent, s, instr.rhs)));
        break;

      // --- Logical ---
      case "And":
        setVar(agent, s, instr.dst, isTruthy(getVar(agent, s, instr.lhs)) && isTruthy(getVar(agent, s, instr.rhs)));
        break;
      case "Or":
        setVar(agent, s, instr.dst, isTruthy(getVar(agent, s, instr.lhs)) || isTruthy(getVar(agent, s, instr.rhs)));
        break;
      case "Not":
        setVar(agent, s, instr.dst, !isTruthy(getVar(agent, s, instr.src)));
        break;

      // --- String/Type ---
      case "Concat":
        setVar(agent, s, instr.dst, valueConcat(getVar(agent, s, instr.lhs), getVar(agent, s, instr.rhs)));
        break;
      case "ToString":
        setVar(agent, s, instr.dst, toDisplayString(getVar(agent, s, instr.src)));
        break;
      case "TypeOf":
        setVar(agent, s, instr.dst, typeName(getVar(agent, s, instr.src)));
        break;
    }
  }
}
